statefulset.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright (C) 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rigging
+
+import (
+	"context"
+	"io"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+// NewStatefulSetControl return new instance of StatefulSet updater
+func NewStatefulSetControl(config StatefulSetConfig) (*StatefulSetControl, error) {
+	err := config.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var ss *appsv1.StatefulSet
+	if config.StatefulSet != nil {
+		ss = config.StatefulSet
+	} else {
+		ss, err = ParseStatefulSet(config.Reader)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	ss.Kind = KindStatefulSet
+	return &StatefulSetControl{
+		StatefulSetConfig: config,
+		statefulSet:       *ss,
+		Entry: log.WithFields(log.Fields{
+			"statefulset": formatMeta(ss.ObjectMeta),
+		}),
+	}, nil
+}
+
+// StatefulSetConfig is a StatefulSet control configuration
+type StatefulSetConfig struct {
+	// Reader with statefulset to update, will be used if present
+	Reader io.Reader
+	// StatefulSet is already parsed statefulset, will be used if present
+	StatefulSet *appsv1.StatefulSet
+	// Client is k8s client
+	Client *kubernetes.Clientset
+}
+
+// CheckAndSetDefaults validates this configuration object and sets defaults
+func (c *StatefulSetConfig) CheckAndSetDefaults() error {
+	var errors []error
+	if c.Reader == nil && c.StatefulSet == nil {
```

Comment:
The message would be simpler if it were just an object reference (`StatefulSet == nil`).
Beside as-is the message is wrong.